### PR TITLE
User deletion and blocking

### LINKED ISSUE #414: User blocking
Api to block a user which then causes NOT_FOUND error when that user tries to access your profile, send a request, events created, etc.

Api to unblock too

### LINKED ISSUE #610: User hiding
Users should be hidden (basically from all queries) if:

* the user is deleted, or
* the user is banned, or
* the user is jailed

This should be done through by filtering with a `hybrid_property` in queries with SQLAlchemy. Similar to the way in which tokens are filtered with `is_valid`:

https://github.com/Couchers-org/couchers/blob/910e9cb0cb8d11d01fca810c816e9e687922dfee/app/backend/src/couchers/models.py#L263-L265

Filtered e.g:

https://github.com/Couchers-org/couchers/blob/66b0ea2ee6211f19e74d52fb0d7ea7788b9782d1/app/backend/src/couchers/servicers/auth.py#L177-L182

This is a global property of a user, separate from user blocking (one user hiding another user, #414)

### LINKED ISSUE #1092: Can get users by email
You shouldn't be able to do this!

https://develop--frontend.preview.couchershq.org/user/[email]

We need to review where `get_user_by_field` is used, and phase it out except for in auth. We need to figure out how the frontend should translate usernames to user IDs as well. In particular, we don't want two trips to the backend for getting a user's profile. At least it needs another function

### PR DESCRIPTION
**Pull request by aapeliv**
Takes over from Sam's #715, #724, #1122, and  #1163, we cleaned this up together. Closes #414, and closes #610.

This adds a new query object that adds some filtering options. For now we grab the blocked users each time, but this can later be cached per-request into `context`, or something.

We also removed `get_user_by_field`, and removed the ability to get a user by email (`GetUser`), so closes #1092.

Also updated most `*_to_pb` style functions to take full `context` instead of just `user_id`.

**Backend checklist**
- [x] Formatted my code by running `isort . && black .` in `app/backend`
- [x] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [x] Run the backend locally and it works
- [x] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

### CODE REVIEW

File: app/pb/blocking.proto [hunk 1]
```diff
@@ -0,0 +1,27 @@
+syntax = "proto3";
+
+package org.couchers.blocking;
+
+import "google/protobuf/empty.proto";
+
+service Blocking {
```
Comment (kalvdans) [inline app/pb/blocking.proto:7]: Describe that you have to be logged in to use this service. I guess we should allow jailed users to update blocking.

File: app/backend/src/tests/test_fixtures.py [hunk 2]
```diff
@@ -95,7 +97,7 @@ def db():
     recreate_database()
 
 
-def generate_user(*_, **kwargs):
+def generate_user(*_, delete=False, **kwargs):
```
Comment (kalvdans) [inline app/backend/src/tests/test_fixtures.py]: Name the flag invisible to better reflect the actual effect. Document it in the docstring.

Also, a drive-by fix is to replace `*_, ` with `*, ` to make sure everyone uses keywords when calling it.

File: app/backend/src/couchers/models.py [hunk 3]
```diff
@@ -239,6 +243,60 @@ def __repr__(self):
         return f"User(id={self.id}, email={self.email}, username={self.username})"
 
 
+# CouchersQuery is placed here (instead of the more natural db) because we import stuff from models into db, so need to
+# avoid circular imports. Additionally, we need it in the relationship configuration, but it requires the User model, so
+# it's as high up as we can get it, and then we patch the User relationships in manually.
```
Comment (kalvdans) [inline app/backend/src/couchers/models.py]: Instead of creating a CouchersQuery just create standalone functions that takes a query as the first argument. Will be a bit different/uglier when using it though, mixing yoda speak with regular speak:

```
                filter_users_column(session.query(Message.id)
                    .join(HostRequest, Message.conversation_id == HostRequest.conversation_id)
                    .outerjoin(
                        message_2, and_(Message.conversation_id == message_2.conversation_id, Message.id < message_2.id)
                    )
                    .filter(HostRequest.from_user_id == context.user_id),
                        context, HostRequest.to_user_id)
                .filter(message_2.id == None)
                .filter(HostRequest.from_last_seen_message_id < Message.id)
                .count()
```

Or introduce a third file to avoid circular imports.

File: app/backend/src/couchers/models.py [hunk 4]
```diff
@@ -1258,3 +1316,37 @@ def ready_for_retry(self):
 
     def __repr__(self):
         return f"BackgroundJob(id={self.id}, job_type={self.job_type}, state={self.state}, next_attempt_after={self.next_attempt_after}, try_count={self.try_count}, failure_info={self.failure_info})"
+
+
+class UserBlocks(Base):
```
Comment (kalvdans) [inline app/backend/src/couchers/models.py]: traditionally, the class name is singular while the table is plural. No need to stick to it though!

File: app/backend/src/couchers/models.py [hunk 5]
```diff
@@ -1258,3 +1316,37 @@ def ready_for_retry(self):
...
+
+    blocking_user_id = Column(ForeignKey("users.id"), nullable=False)
+    blocked_user_id = Column(ForeignKey("users.id"), nullable=False)
+    time_blocked = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
+
+    is_blocking_user = relationship("User", backref="is_blocking_user", foreign_keys="UserBlocks.blocking_user_id")
+    is_blocked_user = relationship("User", backref="is_blocked_user", foreign_keys="UserBlocks.blocked_user_id")
+
+
+def blocked_users(session, user_id):
```
Comment (kalvdans) [inline app/backend/src/couchers/models.py]: Move to db.py?

File: app/backend/src/couchers/models.py [hunk 6]
```diff
@@ -149,6 +150,7 @@ class User(Base):
     additional_information = Column(String, nullable=True)  # CommonMark without images
 
     is_banned = Column(Boolean, nullable=False, default=False)
+    is_deleted = Column(Boolean, nullable=False, default=False)
```
Comment (kalvdans) [inline app/backend/src/couchers/models.py:152]: I think we should name this column `is_invisible` to indicate that it can be resurrected at any time.

File: app/backend/src/couchers/servicers/blocking.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                    blocked_user_id=request.user_id,
+                )
+                session.add(user_block)
+                session.commit()
+
+        return empty_pb2.Empty()
+
+    def UnblockUser(self, request, context):
+        if context.user_id == request.user_id:
+            context.abort(grpc.StatusCode.INVALID_ARGUMENT, errors.CANT_UNBLOCK_SELF)
```
Comment (kalvdans) [inline app/backend/src/couchers/servicers/blocking.py]: Redundant, will be caught by USER_NOT_BLOCKED further down anyway.

File: app/backend/src/couchers/servicers/communities.py [hunk 8]
```diff
@@ -19,7 +19,7 @@
 MAX_PAGINATION_LENGTH = 25
 
 
-def _parents_to_pb(node_id, user_id):
+def _parents_to_pb(node_id, context):
```
Comment (kalvdans) [inline app/backend/src/couchers/servicers/communities.py]: Remove argument instead, it seems to be unused?

**Review by kalvdans — CHANGES_REQUESTED** [review]

File: app/pb/blocking.proto (hunk 1 above)
Comment (aapeliv) [reply app/pb/blocking.proto:7]: Good catch

File: app/backend/src/couchers/models.py (hunk 6 above)
Comment (aapeliv) [reply app/backend/src/couchers/models.py:152]: "invisible" means that it's not visible: but you can be invisible if you've been banned or you can be invisible if you've been deleted. There's the `is_visible` helper function that combines both of these.

I think it's good to keep this as `is_deleted` to clarify the purpose of the flag.

File: app/backend/src/couchers/models.py (hunk 3 above)
Comment (aapeliv) [reply app/backend/src/couchers/models.py]: I really think `CouchersQuery` makes everything *a lot* nicer, cleaner, and more concise than having separate functions that you have to wrap around your query. In some of the bigger queries, they just become a huge mess otherwise, especially if you have to use multiple filters.

> Or introduce a third file to avoid circular imports.

The problem is that:
* `CouchersQuery` requires the `User` class to be defined
* `relationship` requires `CouchersQuery` to be defined
* Basically all classes require `relationship` to be defined

I'm not sure how to get around this, but it'd be really nice to find a cleaner way.

Maybe if we split `User` into another file? The split of `models.py` is long overdue, but those PRs never seem to go through and somehow become stale so the merges are a pain...

File: app/backend/src/couchers/models.py (hunk 5 above)
Comment (aapeliv) [reply app/backend/src/couchers/models.py]: It requires `UserBlock` from this file, so that'd cause circular imports...

File: app/backend/src/couchers/models.py (hunk 6 above)
Comment (kalvdans) [reply app/backend/src/couchers/models.py:152]: Fair enough!

File: app/pb/blocking.proto [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    //Returns all users blocked by a given account
+  }
+}
+
+message BlockUserReq {
+  int64 user_id = 1;
+}
+
+message UnblockUserReq {
+  int64 user_id = 1;
```
Comment (aapeliv) [inline app/pb/blocking.proto]: @kalvdans: I think these `user_id`s should be changed to `string username`.

There's no way for the requester to get the user_id because `GetUser` will return not found for blocked users.

Also if user A blocks user B, then user B might also want to block user A, so it's not guaranteed that a the blocker can find the user id either

Comment (kalvdans) [reply app/pb/blocking.proto]: If I block you, I should still be able to GetUser you IMHO. But indeed if we have a mutual blockage I guess I have to call you and ask which user_id number you have.

File: app/backend/src/couchers/query.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
+from sqlalchemy import or_
+from sqlalchemy.orm import Query, aliased
+
+from couchers.models import User, UserBlock
+from couchers.utils import is_valid_email, is_valid_user_id, is_valid_username
+
+
+class CouchersQuery(Query):
+    def _blocked_users(self, session, user_id):
```
Comment (kalvdans) [inline app/backend/src/couchers/query.py]: move to top level, no need to have it as a member function. It doesn't chain another filter onto the query.

File: app/backend/src/couchers/query.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        # no fields match, this will return no rows
+        return self.filter(False)
+
+    def filter_users(self, context, table=User):
+        """
+        Filters out users that should not be visible: blocked, deleted, or banned
+
+        Filters the given table, assuming it's already joined/selected from
+        """
+        hidden_users = self._blocked_users(self.session, context.user_id)
```
Comment (kalvdans) [inline app/backend/src/couchers/query.py]: This call does a full sql query to the database, it would be better if it could augment the current query with the right filters. But it could be an improvement to be done later.

File: app/pb/blocking.proto (hunk 9 above)
Comment (aapeliv) [reply app/pb/blocking.proto]: I think it's better to just use username here.

Regarding getting the user with GetUser; there should maybe be an option to get some subset of information. But I don't think it's very good that the blocker has the power to see the other user but the other user can't see the blocker, could be abused.

File: app/backend/src/couchers/servicers/communities.py [hunk 12]
```diff
@@ -35,22 +35,52 @@ def _parents_to_pb(node_id, user_id):
...
-        can_moderate = can_moderate_node(session, user_id, node.id)
+        can_moderate = can_moderate_node(session, context.user_id, node.id)
+
+        member_count = (
+            session.query(ClusterSubscription)
+            .filter_users_column(context, ClusterSubscription.user_id)
+            .filter(ClusterSubscription.cluster_id == node.official_cluster.id)
+            .count()
+        )
+        member = (
```
Comment (kalvdans) [inline app/backend/src/couchers/servicers/communities.py]: is_member

File: app/backend/src/couchers/servicers/communities.py [hunk 13]
```diff
@@ -35,22 +35,52 @@ def _parents_to_pb(node_id, user_id):
...
+        )
+
+        admin_count = (
+            session.query(ClusterSubscription)
+            .filter_users_column(context, ClusterSubscription.user_id)
+            .filter(ClusterSubscription.cluster_id == node.official_cluster.id)
+            .filter(ClusterSubscription.role == ClusterRole.admin)
+            .count()
+        )
+        admin = (
```
Comment (kalvdans) [inline app/backend/src/couchers/servicers/communities.py]: is_admin

File: app/backend/src/couchers/utils.py [hunk 14]
```diff
@@ -14,6 +15,47 @@
...
+    """
+    Checks if it has at least one non-whitespace character
+    """
+    return re.match(r"\S+", field) is not None
+
+
+def is_valid_email(field):
+    """
+    From SO
+    """
```
Comment (kalvdans) [inline app/backend/src/couchers/utils.py]: comment instead of docstring

**Review by kalvdans — DISMISSED** [review]
Small remarks, but looks good in general!

File: app/backend/src/couchers/query.py (hunk 11 above)
Comment (aapeliv) [reply app/backend/src/couchers/query.py]: I'll leave it for later

File: app/pb/blocking.proto (hunk 9 above)
Comment (kalvdans) [reply app/pb/blocking.proto]: I guess one can argue either way. Symmetrical blocking is conceptually simpler indeed.

OK, please go with usernames here and in GetBlockedUsers

**Comment by aapeliv** [comment]
All done

**Review by kalvdans — DISMISSED** [review]
LGTM! Please merge when CI is happy!
